### ISSUE TITLE
Enforcing max_norm on entities and operators

### ISSUE THREAD

**Issue by simonepri** [facebookresearch/PyTorch-BigGraph#54]
My understanding is that currently, the `max_norm` option allows users to enforce the norm of the embeddings learned for the entities to be smaller than a specific value.

If that's correct I have to two questions:
* In operators like complex_diagonal where the embedding learned is made of two components, how can one enforce max_norm on the two separate parts? (enforce max_norm on the first half and max_norm on the second half)

* In case there's the need to enforce max_norm also on the parameters learned for the operator, how can one achieve that?

**Comment by lerks** [comment]
The `max_norm` config parameter only enforces a maximum (L2) norm on the embedding vectors of the entities. There's currently no constraint on the operator parameters. There's no special reason for it, as far as I know, it's just that we never saw a strong reason to implement it.

As you mentioned, however, it's not always clear how to enforce such a "max norm" type of constraint. In some cases, like translation or diagonal, there may be an (arguably) natural way because the parameters are a vector of the same size as the embeddings. But already for linear or affine operators it's less obvious, and so also for the complex diagonal.

(As a side note, it may also make sense to have a different "upper bound" for parameters than the one for entity embeddings, just like we have a different learning rate).

If you're interested in implementing this we would probably accept a pull request and we can discuss here the specifics. Since you asked about the complex diagonal operator, I think one way of doing the norm of its parameters would be to do the norm of the D/2-dimensional complex vector, which if I did my math correctly should correspond to doing the regular L2 norm of the same data interpreted as a D-dimensional real vector. But I don't know whether this makes more or less sense than other approaches.

**Comment by simonepri** [comment]
> There's currently no constraint on the operator parameters

I'm thinking about how to implement this.
One way I found, is to provide an additional `options` dictionary to the operator's constructor and then use `clip_grad_norm_` to enforce the max norm on the parameters.

```python3
class TranslationOperator(AbstractOperator):

    def __init__(self, dim: int, options: Dict):
        super().__init__(dim)
        if options.max_norm is not None:
            self.translation = nn.utils.clip_grad_norm_(nn.Parameter(torch.zeros((self.dim,))), options.max_norm)
        else:
            self.translation = nn.Parameter(torch.zeros((self.dim,)))

    def forward(self, embeddings: FloatTensorType) -> FloatTensorType:
        match_shape(embeddings, ..., self.dim)
        return embeddings + self.translation
```

The dictionary would be provided by the users in a way like this:

```python3
relations=[{
  ...
  'operator': 'translation',
  'options': {'max_norm': 1.0}
}]
```

Do you have any better idea?

**Comment by simonepri** [comment]
> There's currently no constraint on the operator parameters

I'm thinking about how to implement this.
One way I found, is to provide an additional `options` dictionary to the operator's constructor and then use the `Embedding` class to enforce the max norm on the parameters. (But I don't like it that much)

```python3
class TranslationOperator(AbstractOperator):

    def __init__(self, dim: int):
        super().__init__(dim)
        self.translation = nn.Embedding(1, self.dim, max_norm=1.0)
        nn.init.constant_(self.translation.weight.data, val=0)

    def forward(self, embeddings: FloatTensorType) -> FloatTensorType:
        match_shape(embeddings, ..., self.dim)
        return embeddings + self.translation(torch.LongTensor([0]))[0]
```

The dictionary would be provided by the users in a way like this:

```python3
relations=[{
  ...
  'operator': 'translation',
  'options': {'max_norm': 1.0}
}]
```

Do you have any better idea?

**Comment by lerks** [comment]
I'm not that convinced about this. Wrt the changes to the operator, I understand why you're using an embedding, and I don't really have better ways of doing this, but it does make the code harder to understand. Wrt the config format, it could work but it looks hard to make compatible with the current parsing code, which requires strict typing (this means that we'd need to have a single type for the values of that dictionary, thus disallowing strings, enums, integers, ...).

While I think about what to do here, I'm also wondering whether a max norm is really the best way to go or if we could investigate other more general approaches. I'm considering whether adding a regularization term to the loss could also achieve the effect of keeping the norms down without such a strict cutoff.

**Comment by simonepri** [comment]
> I'm considering whether adding a regularization term to the loss could also achieve the effect of keeping the norms down without such a strict cutoff.

This is actually what is done in many papers.

From the [TransH paper](https://pdfs.semanticscholar.org/2a3f/862199883ceff5e3c74126f0c80770653e05.pdf):

> The following constraints are considered when we minimize the loss L:
> <img height="100" src="https://user-images.githubusercontent.com/3505087/60045731-dce17e80-96c5-11e9-9de2-42ed57c18fb4.png">
> Instead of directly optimizing the loss function with constraints, we convert it to the following unconstrained loss by means of soft constraints, where C is a hyper-parameter weighting the importance of soft constraints.
> <img height="100" src="https://user-images.githubusercontent.com/3505087/60045594-83794f80-96c5-11e9-8a91-1bf2425c773a.png">

**Comment by lerks** [comment]
At some point we did hack in some regularization code when trying to replicate @timlacroix's ComplEx model but then we removed it because it only gave us a small improvement.

If you want to propose some interface that could be used for "regularizers" to plug into the training loop I can review it. I gave it a try back then but it wasn't trivial to make it both efficient and general enough.

If you want to just experiment with whether constraining operators' parameters helps, you can use the operator class you posted earlier to run some experiments locally and then we'll discuss again how to merge this once we're sure it brings some benefit.

**Comment by rkindi** [comment]
Regularizer support was added in https://github.com/facebookresearch/PyTorch-BigGraph/commit/cb4dccfcc020da598c6a64c6370cd1a121027567. The regularization term can be a function of the operator params:

https://github.com/facebookresearch/PyTorch-BigGraph/blob/67c42f9002ad7d0f03d532ea8cc165818e48bafc/torchbiggraph/regularizers.py#L73-L76
